### ISSUE TITLE
Really want to use this but can't figure out how

### ISSUE THREAD

**Issue by martinmckenna** [longlho/ts-transform-react-intl#29]
I'm using webpack with ts-loader and I'd like to extract my react-intl strings to JSON files.

But I can't figure out how to get this plugin working correctly. I've tried looking at other issues and still can't find any help.

For starters, how do I tell this plugin where to create the JSON file? Also, how am I supposed to be using the `onMsgExtracted` property

This is what I've tried so far:

`webpack.config`

```js
          {
            test: /\.(ts|tsx)$/,
            include: paths.appSrc,
            use: [
              {
                loader: require.resolve('ts-loader'),
                options: {
                  // disable type checker - we will use it in fork plugin
                  transpileOnly: true,
                  getCustomTransformers: () => ({
                    before: [
                      intl.transform({
                        idPrefix: "prefix",
                        onMsgExtracted: intl.aggregate({}),
                      }),
                    ]
                  })
                },
              },
            ],
          },
```

`MyComponent.tsx`

```js
import { FormattedMessage } from 'react-intl';

export class SummaryPanel extends React.Component<Props, {}> {
  render() {
    return (
      <React.Fragment>
        <FormattedMessage
          defaultMessage="Contact Information"
          id="ContactInfo.heading"
        />
      </React.Fragment>
    );
  }
}
```

**Comment by longlho** [comment]
I'll try to whip up an example tonight or tomorrow. To unblock you right now, for `intl.aggregrate(msgs)`, `msgs` is just a result object you declare and everytime it sees an translation it'll modify the object inline.

The tricky part is to write out that result object to a file once webpack's done.

**Comment by martinmckenna** [comment]
@longlho ah okay. So aggregate() works similiarly to array.reduce in that you pass it a accumulator? Or do you mean something different by “result object?”

Also thank you for working on an example. I’m very much looking to accomplish the functionality of babel-plugin-react-intl, and am hoping this lib will make things easier

**Comment by longlho** [comment]
Yes it works like an accumulator

**Comment by martinmckenna** [comment]
Been trying some other attempts. Most recent failed attempt was to abstract the transformer into my own file:

`webpack.config.js`

```js
              {
                loader: require.resolve('ts-loader'),
                options: {
                  // disable type checker - we will use it in fork plugin
                  transpileOnly: true,
                  getCustomTransformers: () => ({
                    before: [mergeMessages]
                  })
                },
              },
```

`merge_messages.js`

```js
var fs = require("fs");
var ts = require('typescript');
var mkdirpSync = require("mkdirp").sync;
var intl = require('ts-transform-react-intl');

const outputDir = './build/';

var transformer = function (context) {
  console.log('running script');

  // aggregate all messages into one big object
  var messages = {};

  var intlTransformer = intl.transform({
    idPrefix: "prefix",
    onMsgExtracted: intl.aggregate(messages),
  });

  intlTransformer();

  console.log(messages);

  // Create a new directory that we want to write the aggregate messages to
  mkdirpSync(outputDir);
  // Write the messages to this directory
  fs.writeFileSync('data.json', `{ "en": ${JSON.stringify(messages, null, 2)} }`);
}

exports["default"] = transformer;
```

But the build errors out with `Module build failed: TypeError: f is not a function`

**Comment by longlho** [comment]
yeah that's not gonna work. `onMsgExtracted` gets called on every file. What you need is at the end of the whole webpack flow, write the accumulated map out.

I initially did this w/ typescript compiler wrapper because it's synchronous & just run through the whole project. `webpack` is not synchronous so you'd almost need a hook into its lifecycle to know when it's done to write the file out. This is not an easy task (similar to `https://github.com/webpack-contrib/extract-text-webpack-plugin` where all CSS gets extracted to a file).

Some pointer would be https://webpack.js.org/api/compilation-hooks/ or https://webpack.js.org/api/plugins/#custom-hooks

The flow for webpack would be something like:
```
// Declare this out of scope
const msgs = {}

webpack({
...
{
                loader: require.resolve('ts-loader'),
                options: {
                  // disable type checker - we will use it in fork plugin
                  transpileOnly: true,
                  getCustomTransformers: () => ({
                    before: [mergeMessages]
                  })
                },
              },
plugins: {
...<something that hooks on done & output msgs to a JSON file>
}
})
```

**Comment by longlho** [comment]
Oh looks like `webpack` takes in a callback `webpack(config, doneCb)` so u can output in `doneCb` maybe?

**Comment by martinmckenna** [comment]
those are some good thoughts. I'll try creating a plugin that can hook into after compile tonight

**Comment by martinmckenna** [comment]
Looks like hooking into the callbacks didn't work either.

The messages object is always empty

`GenerateMessagesFile.js`

```js
'use strict';
var fs = require('fs');
var mkdirp = require('mkdirp');

const outputPath = '../build/messages/';

class GenerateMessageFile {
  constructor(messageObj) {
    this.messageObj = messageObj;
  };

  apply(compiler) {
    compiler.plugin('done', (compilation) => {
      console.log(this.messageObj);

      // Create a new directory that we want to write the aggregate messages to
      mkdirp.sync(outputPath);

      // Write the messages to this directory
      fs.writeFileSync(outputPath + 'messages.json', `{ "en": ${JSON.stringify(this.messageObj, null, 2)} }`);

    })
  }
}

module.exports = GenerateMessageFile;
```

`webpack.config.js`

```js 
          {
            test: /\.(ts|tsx)$/,
            include: paths.appSrc,
            use: [
              {
                loader: require.resolve('ts-loader'),
                options: {
                  // disable type checker - we will use it in fork plugin
                  transpileOnly: true,
                  getCustomTransformers: () => ({
                    before: [
                      intl.transform({
                        idPrefix: "prefix",
                        onMsgExtracted: intl.aggregate(messages),
                      }),
                    ]
                  })
                },
              },
            ],
          },

  plugins: [
    new GenerateMessageFile(messages),
 ]

```

**Comment by martinmckenna** [comment]
Ah okay so it seems like this plugin will not pick up the text if you’re solely using the react-intl React Component FormattedMessage and that you need to use the API function.

The issue i’m running into now is that the out-of-scope messages var is undefined by the time my webpack hook runs

Would you mind if i just made a PR to allow for passing in a file path to write to? I think that functionality would be widely useful

**Comment by longlho** [comment]
Yeah PR’s welcome :)
